DataStructures/BinarySearchTree/BinarySearchTree.cs

```diff
@@ -0,0 +1,400 @@
+        /// Thrown if key is already in the BST.
+        /// </exception>
+        private void Add(BinarySearchTreeNode<TKey> node, TKey key)
+        {
+            int compareResult = node.Key.CompareTo(key);
+            if (compareResult > 0)
+            {
+                if (node.Left != null)
+                {
+                    Add(node.Left, key);
+                }
+                else
+                {
+                    var newNode = new BinarySearchTreeNode<TKey>(key);
+                    node.Left = newNode;
+                }
+            }
+            else if (compareResult < 0)
+            {
+                if (node.Right != null)
+                {
+                    Add(node.Right, key);
+                }
+                else
+                {
+                    var newNode = new BinarySearchTreeNode<TKey>(key);
+                    node.Right = newNode;
+                }
+            }
+
+            // Key is already in tree.
+            else
+            {
+                throw new ArgumentException($"Key \"{key}\" already exists in tree!");
+            }
+        }
+
+        /// <summary>
+        /// Removes a node with the specified key from the BST.
+        /// </summary>
+        /// <param name="parent">The parent node of <paramref name="node"/>.</param>
+        /// <param name="node">The node to check/search from.</param>
+        /// <param name="key">The key to remove.</param>
+        /// <returns>true if the operation was successful, false otherwise.</returns>
+        /// <remarks>
+        /// Removing a node from the BST can be split into three cases:
+        /// <br></br>
+        /// 0. The node to be removed has no children.In this case, the node can just be removed from the tree.
+        /// <br></br>
+        /// 1. The node to be removed has one child. In this case, the node's child is moved to the node's parent,
+        /// then the node is removed from the tree.
+        /// <br></br>
+        /// 2. The node to be removed has two children. In this case, we must find a suitable node among the children
+        /// subtrees to replace the node. This can be done with either the in-order predecessor or the in-order successor.
+        /// The in-order predecessor is the largest node in Left subtree, or the largest node that is still smaller then the
+        /// current node. The in-order successor is the smallest node in the Right subtree, or the smallest node that is
+        /// still larger than the current node. Either way, once this suitable node is found, remove it from the tree (it
+        /// should be either a case 1 or 2 node) and replace the node to be removed with this suitable node.
+        /// <br></br>
+        /// More information: https://en.wikipedia.org/wiki/Binary_search_tree#Deletion .
+        /// </remarks>
+        private bool Remove(BinarySearchTreeNode<TKey> parent, BinarySearchTreeNode<TKey>? node, TKey key)
```

Comment:
Codacy found an issue: [Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277689)